core/trino-main/src/main/java/io/trino/operator/TopNProcessor.java

```diff
@@ -76,30 +87,46 @@ public void addInput(Page page)
 
     public Page getOutput()
     {
-        if (outputIterator == null) {
+        if (outputIterator == null && topNBuilder != null) {
             // start flushing
             outputIterator = topNBuilder.buildResult();
         }
 
-        Page output = null;
-        if (outputIterator.hasNext()) {
+        Page output;
+        if (outputIterator != null && outputIterator.hasNext()) {
             output = outputIterator.next();
         }
         else {
-            outputIterator = emptyIterator();
+            closeTopNBuilder();
+            return null;
         }
         updateMemoryReservation();
         return output;
     }
 
     public boolean noMoreOutput()
     {
-        return outputIterator != null && !outputIterator.hasNext();
+        return outputIterator == null;
     }
 
     private void updateMemoryReservation()
     {
-        requireNonNull(topNBuilder, "topNBuilder is null");
-        localUserMemoryContext.setBytes(topNBuilder.getEstimatedSizeInBytes());
+        if (topNBuilder == null) {
+            localMemoryContext.setBytes(0);
+            return;
+        }
+        localMemoryContext.setBytes(topNBuilder.getEstimatedSizeInBytes());
+    }
+
+    public boolean isBuilderFull()
+    {
+        return topNBuilder != null && topNBuilder.getEstimatedSizeInBytes() >= maxFlushableBytes;
+    }
```

Comment:
The concept of flushable bytes and partials complicates this class more than it needs to. Flushable thresholds and partials is a concept that the operator should maintain. At this level, we maybe should just expose a getEstimatedSizeInBytes() method:

```java
    public long getEstimatedSizeInBytes()
    {
        return topNBuilder == null ? 0 : topNBuilder.getEstimatedSizeInBytes();
    }
```
Just let the caller track what to do with this value.

Reply:
Done